File: server/soul/catalog.go
package soul

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"regexp"
	"sort"

	"cloud-clicker/server/decimal"
)

var ErrInvalidCatalog = errors.New("invalid soul catalog")

type BandMember string

const (
	BandWhole    BandMember = "whole"
	BandDimming  BandMember = "dimming"
	BandHollow   BandMember = "hollow"
	BandNearZero BandMember = "near_zero"
)

type OwnerKind string

const (
	OwnerEvent     OwnerKind = "event"
	OwnerLongevity OwnerKind = "longevity"
	OwnerContract  OwnerKind = "contract"
	OwnerFixture   OwnerKind = "fixture"
)

type EndingVariant string

const (
	EndingEarnestAscension EndingVariant = "earnest_ascension"
	EndingTrainingData     EndingVariant = "training_data"
)

type Policy struct {
	Floor   int64 `json:"soul_floor"`
	Initial int64 `json:"soul_initial"`
	Max     int64 `json:"soul_max"`
}

type Band struct {
	Member             BandMember `json:"band_member"`
	MinInclusive       int64      `json:"min_inclusive"`
	MaxInclusive       int64      `json:"max_inclusive"`
	HumanContentLocked bool       `json:"human_content_locked"`
	ReasonKey          string     `json:"reason_key"`
}

type DebitSource struct {
	SourceID       string    `json:"source_id"`
	OwnerKind      OwnerKind `json:"owner_kind"`
	Amount         int64     `json:"amount"`
	MayExhaust     bool      `json:"may_exhaust"`
	SingleUse      bool      `json:"single_use"`
	CurtainCopyKey string    `json:"curtain_copy_key"`
}

type RecoveryActivity struct {
	ActivityID         string `json:"activity_id"`
	DurationAttendedMS int64  `json:"duration_attended_ms"`
	RecoveryAmount     int64  `json:"recovery_amount"`
	ReasonKey          string `json:"reason_key"`
}

type EndingPolicy struct {
	WholeVariant    EndingVariant `json:"whole_variant"`
	DepletedVariant EndingVariant `json:"depleted_variant"`
}

type Catalog struct {
	SchemaVersion      int                `json:"schema_version"`
	Policy             Policy             `json:"policy"`
	Bands              []Band             `json:"bands"`
	DebitSources       []DebitSource      `json:"debit_sources"`
	RecoveryActivities []RecoveryActivity `json:"recovery_activities"`
	EndingPolicy       EndingPolicy       `json:"ending_policy"`
}

type Declarations struct {
	CopyKeys    map[string]struct{}
	EpochSeeded bool
}

var mechanicalID = regexp.MustCompile(`^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*$`)

func LoadCatalog(data []byte, declarations Declarations) (*Catalog, error) {
	if !uniqueJSONKeys(data) || !exactSoulCatalogKeys(data) || len(declarations.CopyKeys) == 0 {
		return nil, ErrInvalidCatalog
	}
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	var catalog Catalog
	if decoder.Decode(&catalog) != nil || catalog.SchemaVersion != 1 {
		return nil, ErrInvalidCatalog
	}
	var trailing any
	if !errors.Is(decoder.Decode(&trailing), io.EOF) || validateCatalog(&catalog, declarations) != nil {
		return nil, ErrInvalidCatalog
	}
	return &catalog, nil
}

func exactSoulCatalogKeys(data []byte) bool {
	if !exactKeys(data, "bands", "debit_sources", "ending_policy", "policy", "recovery_activities", "schema_version") {
		return false
	}
	var root map[string]json.RawMessage
	if json.Unmarshal(data, &root) != nil || !exactKeys(root["policy"], "soul_floor", "soul_initial", "soul_max") ||
		!exactKeys(root["ending_policy"], "depleted_variant", "whole_variant") {
		return false
	}
	return exactRows(root["bands"], "band_member", "human_content_locked", "max_inclusive", "min_inclusive", "reason_key") &&
		exactRows(root["debit_sources"], "amount", "curtain_copy_key", "may_exhaust", "owner_kind", "single_use", "source_id") &&
		exactRows(root["recovery_activities"], "activity_id", "duration_attended_ms", "reason_key", "recovery_amount")
}

func exactRows(data []byte, keys ...string) bool {
	var rows []json.RawMessage
	if json.Unmarshal(data, &rows) != nil || rows == nil {
		return false
	}
	for _, row := range rows {
		if !exactKeys(row, keys...) {
			return false
		}
	}
	return true
}

func validateCatalog(catalog *Catalog, declarations Declarations) error {
	if catalog.Policy.Floor < 0 || catalog.Policy.Max > decimal.MaxExactInteger || catalog.Policy.Floor > catalog.Policy.Initial || catalog.Policy.Initial > catalog.Policy.Max ||
		len(catalog.Bands) != 4 || catalog.DebitSources == nil || catalog.RecoveryActivities == nil ||
		catalog.EndingPolicy.WholeVariant != EndingEarnestAscension || catalog.EndingPolicy.DepletedVariant != EndingTrainingData {
		return ErrInvalidCatalog
	}
	wantMembers := [...]BandMember{BandNearZero, BandHollow, BandDimming, BandWhole}
	next := catalog.Policy.Floor
	for index, band := range catalog.Bands {
		if band.Member != wantMembers[index] || band.MinInclusive != next || band.MinInclusive > band.MaxInclusive ||
			band.MaxInclusive > catalog.Policy.Max || band.HumanContentLocked != (band.Member == BandNearZero) || !copyKey(declarations, band.ReasonKey) {
			return ErrInvalidCatalog
		}
		if band.MaxInclusive == decimal.MaxExactInteger {
			next = band.MaxInclusive
		} else {
			next = band.MaxInclusive + 1
		}
	}
	if catalog.Bands[len(catalog.Bands)-1].MaxInclusive != catalog.Policy.Max {
		return ErrInvalidCatalog
	}
	prior := ""
	for _, source := range catalog.DebitSources {
		if !mechanicalID.MatchString(source.SourceID) || source.SourceID <= prior || !validOwner(source.OwnerKind) || source.Amount < 1 || source.Amount > decimal.MaxExactInteger ||
			source.MayExhaust != source.SingleUse || !copyKey(declarations, source.CurtainCopyKey) || declarations.EpochSeeded && source.OwnerKind == OwnerFixture {
			return ErrInvalidCatalog
		}
		prior = source.SourceID
	}
	prior = ""
	for _, activity := range catalog.RecoveryActivities {
		if !mechanicalID.MatchString(activity.ActivityID) || activity.ActivityID <= prior || activity.DurationAttendedMS < 1 || activity.DurationAttendedMS > decimal.MaxExactInteger ||
			activity.RecoveryAmount < 1 || activity.RecoveryAmount > decimal.MaxExactInteger || !copyKey(declarations, activity.ReasonKey) {
			return ErrInvalidCatalog
		}
		prior = activity.ActivityID
	}
	if !declarations.EpochSeeded && (len(catalog.DebitSources) == 0 || len(catalog.RecoveryActivities) == 0) {
		return ErrInvalidCatalog
	}
	return nil
}

func validOwner(value OwnerKind) bool {
	return value == OwnerEvent || value == OwnerLongevity || value == OwnerContract || value == OwnerFixture
}

func copyKey(declarations Declarations, value string) bool {
	_, ok := declarations.CopyKeys[value]
	return ok
}

func (catalog *Catalog) BandFor(value int64) (Band, bool) {
	if catalog == nil || value < catalog.Policy.Floor || value > catalog.Policy.Max {
		return Band{}, false
	}
	index := sort.Search(len(catalog.Bands), func(index int) bool { return catalog.Bands[index].MaxInclusive >= value })
	if index == len(catalog.Bands) {
		return Band{}, false
	}
	return catalog.Bands[index], true
}

func (catalog *Catalog) HumanContentLocked(value int64) (bool, error) {
	band, ok := catalog.BandFor(value)
	if !ok {
		return false, ErrInvalidCatalog
	}
	return band.HumanContentLocked, nil
}

func (catalog *Catalog) DebitSource(id string) (DebitSource, bool) {
	index := sort.Search(len(catalog.DebitSources), func(index int) bool { return catalog.DebitSources[index].SourceID >= id })
	if index == len(catalog.DebitSources) || catalog.DebitSources[index].SourceID != id {
		return DebitSource{}, false
	}
	return catalog.DebitSources[index], true
}

func (catalog *Catalog) RecoveryActivity(id string) (RecoveryActivity, bool) {
	index := sort.Search(len(catalog.RecoveryActivities), func(index int) bool { return catalog.RecoveryActivities[index].ActivityID >= id })
	if index == len(catalog.RecoveryActivities) || catalog.RecoveryActivities[index].ActivityID != id {
		return RecoveryActivity{}, false
	}
	return catalog.RecoveryActivities[index], true
}

func exactKeys(data []byte, keys ...string) bool {
	var object map[string]json.RawMessage
	if json.Unmarshal(data, &object) != nil || len(object) != len(keys) {
		return false
	}
	for _, key := range keys {
		if _, ok := object[key]; !ok {
			return false
		}
	}
	return true
}

func uniqueJSONKeys(data []byte) bool {
	decoder := json.NewDecoder(bytes.NewReader(data))
	var visit func() bool
	visit = func() bool {
		token, err := decoder.Token()
		if err != nil {
			return false
		}
		delimiter, ok := token.(json.Delim)
		if !ok {
			return true
		}
		switch delimiter {
		case '{':
			seen := map[string]bool{}
			for decoder.More() {
				keyToken, err := decoder.Token()
				key, ok := keyToken.(string)
				if err != nil || !ok || seen[key] {
					return false
				}
				seen[key] = true
				if !visit() {
					return false
				}
			}
			end, err := decoder.Token()
			return err == nil && end == json.Delim('}')
		case '[':
			for decoder.More() {
				if !visit() {
					return false
				}
			}
			end, err := decoder.Token()
			return err == nil && end == json.Delim(']')
		default:
			return false
		}
	}
	return visit()
}

File: client/src/soul/catalog.ts
import { MAX_EXACT_INTEGER } from "../numeric";

export const SOUL_BANDS = ["near_zero", "hollow", "dimming", "whole"] as const;
export type SoulBand = (typeof SOUL_BANDS)[number];
export type SoulOwnerKind = "event" | "longevity" | "contract" | "fixture";
export type SoulEndingVariant = "earnest_ascension" | "training_data";

export interface SoulPolicy { readonly soul_floor: number; readonly soul_initial: number; readonly soul_max: number }
export interface SoulBandRow { readonly band_member: SoulBand; readonly min_inclusive: number; readonly max_inclusive: number; readonly human_content_locked: boolean; readonly reason_key: string }
export interface SoulDebitSource { readonly source_id: string; readonly owner_kind: SoulOwnerKind; readonly amount: number; readonly may_exhaust: boolean; readonly single_use: boolean; readonly curtain_copy_key: string }
export interface SoulRecoveryActivity { readonly activity_id: string; readonly duration_attended_ms: number; readonly recovery_amount: number; readonly reason_key: string }
export interface SoulCatalog {
  readonly schema_version: 1;
  readonly policy: SoulPolicy;
  readonly bands: readonly SoulBandRow[];
  readonly debit_sources: readonly SoulDebitSource[];
  readonly recovery_activities: readonly SoulRecoveryActivity[];
  readonly ending_policy: { readonly whole_variant: "earnest_ascension"; readonly depleted_variant: "training_data" };
}

export interface SoulDeclarations { readonly copyKeys: ReadonlySet<string>; readonly epochSeeded: boolean }

const mechanical = /^[a-z][a-z0-9_]*(?:\.[a-z][a-z0-9_]*)*$/;
const owners = new Set<SoulOwnerKind>(["event", "longevity", "contract", "fixture"]);

export function parseSoulCatalog(source: unknown, declarations: SoulDeclarations): SoulCatalog {
  if (declarations.copyKeys.size === 0) throw new SyntaxError("Soul copy registry is empty");
  const root = exactObject(source, ["schema_version", "policy", "bands", "debit_sources", "recovery_activities", "ending_policy"], "Soul catalog");
  if (root.schema_version !== 1) throw new SyntaxError("invalid Soul catalog version");
  const policyRow = exactObject(root.policy, ["soul_floor", "soul_initial", "soul_max"], "Soul policy");
  const floor = safeInteger(policyRow.soul_floor, 0, MAX_EXACT_INTEGER), maximum = safeInteger(policyRow.soul_max, floor, MAX_EXACT_INTEGER);
  const policy: SoulPolicy = Object.freeze({ soul_floor: floor, soul_initial: safeInteger(policyRow.soul_initial, floor, maximum), soul_max: maximum });
  if (!Array.isArray(root.bands) || root.bands.length !== SOUL_BANDS.length) throw new SyntaxError("Soul bands must be complete");
  let next = floor;
  const bands = Object.freeze(root.bands.map((item, index) => {
    const row = exactObject(item, ["band_member", "min_inclusive", "max_inclusive", "human_content_locked", "reason_key"], "Soul band");
    if (row.band_member !== SOUL_BANDS[index] || row.min_inclusive !== next || typeof row.human_content_locked !== "boolean" || row.human_content_locked !== (row.band_member === "near_zero")) throw new SyntaxError("invalid Soul band partition");
    const max = safeInteger(row.max_inclusive, next, maximum), reason = copyKey(row.reason_key, declarations);
    next = max + 1;
    return Object.freeze({ band_member: row.band_member as SoulBand, min_inclusive: row.min_inclusive as number, max_inclusive: max, human_content_locked: row.human_content_locked, reason_key: reason });
  }));
  if (bands.at(-1)?.max_inclusive !== maximum) throw new SyntaxError("Soul bands do not cover policy domain");
  if (!Array.isArray(root.debit_sources) || !Array.isArray(root.recovery_activities)) throw new SyntaxError("Soul content rows must be arrays");
  let prior = "";
  const debitSources = Object.freeze(root.debit_sources.map((item) => {
    const row = exactObject(item, ["source_id", "owner_kind", "amount", "may_exhaust", "single_use", "curtain_copy_key"], "Soul debit source");
    const sourceId = identifier(row.source_id, "Soul source");
    if (byteCompare(prior, sourceId) >= 0 || typeof row.owner_kind !== "string" || !owners.has(row.owner_kind as SoulOwnerKind) || typeof row.may_exhaust !== "boolean" || row.may_exhaust !== row.single_use || declarations.epochSeeded && row.owner_kind === "fixture") throw new SyntaxError("invalid Soul debit source");
    prior = sourceId;
    return Object.freeze({ source_id: sourceId, owner_kind: row.owner_kind as SoulOwnerKind, amount: safeInteger(row.amount, 1, MAX_EXACT_INTEGER), may_exhaust: row.may_exhaust, single_use: row.single_use as boolean, curtain_copy_key: copyKey(row.curtain_copy_key, declarations) });
  }));
  prior = "";
  const activities = Object.freeze(root.recovery_activities.map((item) => {
    const row = exactObject(item, ["activity_id", "duration_attended_ms", "recovery_amount", "reason_key"], "Soul recovery activity");
    const activityId = identifier(row.activity_id, "Soul activity");
    if (byteCompare(prior, activityId) >= 0) throw new SyntaxError("Soul activities are not byte sorted");
    prior = activityId;
    return Object.freeze({ activity_id: activityId, duration_attended_ms: safeInteger(row.duration_attended_ms, 1, MAX_EXACT_INTEGER), recovery_amount: safeInteger(row.recovery_amount, 1, MAX_EXACT_INTEGER), reason_key: copyKey(row.reason_key, declarations) });
  }));
  if (!declarations.epochSeeded && (debitSources.length === 0 || activities.length === 0)) throw new SyntaxError("fixture Soul catalog needs source and activity rows");
  const ending = exactObject(root.ending_policy, ["whole_variant", "depleted_variant"], "Soul ending policy");
  if (ending.whole_variant !== "earnest_ascension" || ending.depleted_variant !== "training_data") throw new SyntaxError("invalid Soul ending policy");
  return Object.freeze({ schema_version: 1, policy, bands, debit_sources: debitSources, recovery_activities: activities,
    ending_policy: Object.freeze({ whole_variant: "earnest_ascension", depleted_variant: "training_data" }) });
}

export function soulBand(catalog: SoulCatalog, value: number): SoulBandRow {
  safeInteger(value, catalog.policy.soul_floor, catalog.policy.soul_max);
  const band = catalog.bands.find((row) => value <= row.max_inclusive);
  if (!band) throw new RangeError("Soul value outside catalog bands");
  return band;
}

export function humanContentLocked(catalog: SoulCatalog, value: number): boolean { return soulBand(catalog, value).human_content_locked }

function copyKey(value: unknown, declarations: SoulDeclarations): string {
  const result = identifier(value, "Soul copy key");
  if (!declarations.copyKeys.has(result)) throw new SyntaxError("unknown Soul copy key");
  return result;
}

function identifier(value: unknown, label: string): string {
  if (typeof value !== "string" || !mechanical.test(value)) throw new SyntaxError(`invalid ${label}`);
  return value;
}

function safeInteger(value: unknown, minimum: number, maximum: number): number {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value < minimum || value > maximum) throw new SyntaxError("Soul integer outside exact domain");
  return value;
}

function exactObject(source: unknown, keys: readonly string[], label: string): Record<string, unknown> {
  if (source === null || typeof source !== "object" || Array.isArray(source)) throw new SyntaxError(`${label} must be an object`);
  const object = source as Record<string, unknown>, actual = Object.keys(object).sort(byteCompare), expected = [...keys].sort(byteCompare);
  if (actual.length !== expected.length || actual.some((key, index) => key !== expected[index])) throw new SyntaxError(`${label} fields are not exact`);
  return object;
}

function byteCompare(left: string, right: string): number {
  const a = new TextEncoder().encode(left), b = new TextEncoder().encode(right);
  for (let index = 0; index < Math.min(a.length, b.length); index++) if (a[index] !== b[index]) return a[index]! - b[index]!;
  return a.length - b.length;
}
